NorthstarDLL/server/servernethooks.cpp

```diff
@@ -0,0 +1,208 @@
+		goto cleanup;
+	}
+
+	if ((status = BCryptFinishHash(h, (PUCHAR)&hash, (ULONG)sizeof(hash), 0)))
+	{
+		spdlog::error("failed to verify HMAC-SHA256: BCryptFinishHash: error 0x{:08X}", (ULONG)status);
+		goto cleanup;
+	}
+
+	if (std::string(hash, sizeof(hash)) == sig)
+	{
+		result = true;
+		goto cleanup;
+	}
+
+cleanup:
+	if (h)
+		BCryptDestroyHash(h);
+	return result;
+}
+
+static void ProcessAtlasConnectionlessPacketSigreq1(R2::netpacket_t* packet, bool dbg, std::string pType, std::string pData)
+{
+	if (pData.length() < HMACSHA256_LEN)
+	{
+		if (dbg)
+			spdlog::warn("ignoring Atlas connectionless packet (size={} type={}): invalid: too short for signature", packet->size, pType);
+		return;
+	}
+
+	std::string pSig; // is binary data, not actually an ASCII string
+	pSig = pData.substr(0, HMACSHA256_LEN);
+	pData = pData.substr(HMACSHA256_LEN);
+
+	if (!g_pMasterServerManager || !g_pMasterServerManager->m_sOwnServerAuthToken[0])
+	{
+		if (dbg)
+			spdlog::warn(
+				"ignoring Atlas connectionless packet (size={} type={}): invalid (data={}): no masterserver token yet",
+				packet->size,
+				pType,
+				pData);
+		return;
+	}
+
+	if (!VerifyHMACSHA256(std::string(g_pMasterServerManager->m_sOwnServerAuthToken), pSig, pData))
+	{
+		if (!Cvar_net_debug_atlas_packet_insecure->GetBool())
+		{
+			if (dbg)
+				spdlog::warn(
+					"ignoring Atlas connectionless packet (size={} type={}): invalid: invalid signature (key={})",
+					packet->size,
+					pType,
+					std::string(g_pMasterServerManager->m_sOwnServerAuthToken));
+			return;
+		}
+		spdlog::warn(
+			"processing Atlas connectionless packet (size={} type={}) with invalid signature due to net_debug_atlas_packet_insecure",
+			packet->size,
+			pType);
+	}
+
+	if (dbg)
+		spdlog::info("got Atlas connectionless packet (size={} type={} data={})", packet->size, pType, pData);
+
+	std::thread t(&MasterServerManager::ProcessConnectionlessPacketSigreq1, g_pMasterServerManager, pData);
+	t.detach();
+
+	return;
+}
+
+static void ProcessAtlasConnectionlessPacket(R2::netpacket_t* packet)
+{
+	bool dbg = Cvar_net_debug_atlas_packet->GetBool();
+
+	// extract kind, null-terminated type, data
+	std::string pType, pData;
+	for (int i = 5; i < packet->size; i++)
+	{
+		if (packet->data[i] == '\x00')
+		{
+			pType.assign((char*)(&packet->data[5]), (size_t)(i - 5));
+			if (i + 1 < packet->size)
+				pData.assign((char*)(&packet->data[i + 1]), (size_t)(packet->size - i - 1));
+			break;
+		}
+	}
+
+	// v1 HMACSHA256-signed masterserver request
+	if (pType == "sigreq1")
+	{
+		ProcessAtlasConnectionlessPacketSigreq1(packet, dbg, pType, pData);
+		return;
+	}
+
+	if (dbg)
+		spdlog::warn("ignoring Atlas connectionless packet (size={} type={}): unknown type", packet->size, pType);
+	return;
+}
+
+AUTOHOOK(ProcessConnectionlessPacket, engine.dll + 0x117800, bool, , (void* a1, R2::netpacket_t* packet))
+{
+	if (4 < packet->size && packet->data[4] == 'T')
```

Comment:
@BobTheBob9, is `T` acceptable?

Reply:
don't see why not? might be better to have these as named consts in code rather than just magic chars though

Reply:
It is only used in one place, so I do think it's a bit more explicit to keep it here. Will add a comment though.